### ISSUE TITLE
Wrong gradient for `tf.math.xlog1py` when x is `0`

### ISSUE THREAD

**Issue by VictoriaGriffith** [tensorflow/tensorflow#56873]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

tf 2.9

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 20.04

### Python version

3.9

### Current Behaviour?

```shell
The gradient of `tf.math.xlog1py` at `x=0` should be `log1p(y)`. In the example below, `y=1.0`, so the gradient should be `0.6931472`. However, the output gradient is `0.0`, which is wrong!
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
x = tf.zeros([1])
y = 1.0
print("x = ", x)
print("y = ", y)

with tf.GradientTape() as g:
  g.watch(x)
  res = tf.math.xlog1py(x, y)
print("gradient: ", g.gradient(res, x))
theoretical, numerical = tf.test.compute_gradient(lambda x: tf.math.xlog1py(x, y), [x])
print("theoretical: ", theoretical)
print("numerical: ", numerical)
print("log1p(y): ", tf.math.log1p(y))
```

### Relevant log output

```shell
x =  tf.Tensor([0.], shape=(1,), dtype=float32)
y =  1.0
gradient:  tf.Tensor([0.], shape=(1,), dtype=float32)
theoretical:  (array([[0.]], dtype=float32),)
numerical:  (array([[0.6931472]], dtype=float32),)
log1p(y):  tf.Tensor(0.6931472, shape=(), dtype=float32)
```
</details>

**Comment by tilakrayal** [comment]
@chunduriv,
I was able to reproduce the issue on tensorflow v2.8, v3.9 and nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/6a5f44bf0e80f6485a22d26610e93cbe/56873.ipynb).

**Comment by cantonios** [comment]
The issue here is the special definition of `xlog1py`, which is:
```
0,                      if x == 0,
x * log1p(y),    otherwise 
```
So for a 0 input, the gradient is technically 0 (since we use the constant 0 function), and we deliberately set this in the gradient here: https://github.com/tensorflow/tensorflow/blob/039156945c876a3e95e8c8e072c59ed24623b750/tensorflow/python/ops/math_grad.py#L731

Otherwise, if `y` is <= -1 or `Inf` or `NaN`, the result of the gradient would be garbage.

What is your use-case?  If you don't need to handle bad inputs, could you use `x * tf.log1p(y)` explicitly?

**Comment by VictoriaGriffith** [comment]
I understand that when `x` is `0`, the result of `xlogpy` is always `0`. However, since we are computing the gradient w.r.t. `x`, we are fixing the value of `y` to be `1.0`, in this case, `xlogpy` is a linear function of `x` (and is defined for all real values of `x`), and the gradient should be `log1p(y)`, not `0`.

**Comment by VictoriaGriffith** [comment]
Thanks for your answer @cantonios ! Can you also check this open issue for `torch.special.xlog1py` https://github.com/pytorch/pytorch/issues/80770? This bug has been confirmed and PyTorch community are working actively to fix it. So I guess both TensorFlow and PyTorch are consistently wrong on this special case.
